### PR TITLE
Reimplementation of Light source dialog

### LINKED ISSUE #15793: Preferences: Headlight direction changes when page is opened, resetting does not reset to default value
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

Opening the preference page Display > Light sources and clicking ok without changing anything, changes the headlight direction.

A reset of that page does not reset it to the default (0,0,-1) properly.
On a fresh install it is 0,0,-1. When you just open the preferences and go to the Display>Light source tab the headlight direction changes to something to the top left side. When resetting the preferences it does not change.

It seems like the headlight is controlled by either HeadlightDirection or the vector of HeadlightRotationW, X, Y, Z,... it seems the light adjusts to the parameter which was updated last, either direction or the rotation values. Not sure if this can be fixed for 1.0

### Full version info

```shell
OS: Windows 11 build 22631
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.38385 (Git)
Build type: Release
Branch: main
Hash: 58241e027e7187a1ecaeeb75668ef9a497437c68
Python 3.11.9, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: German/Germany (de_DE)
Installed mods: 
  * CfdOF 1.27.4
  * Curves 0.6.42
  * dodo 1.0.1
  * fasteners 0.5.25
  * fasteners.backup1721936159.7542229 0.5.24 (Disabled)
  * freecad.gears 1.2.0
  * Manipulator 1.5.7
  * OpenTheme 2024.7.30
  * OpticsWorkbench 1.0.19
  * Rocket 3.3.0
  * sheetmetal 0.4.23
```

### Subproject(s) affected?

Core

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by xtemp09**
@maxwxyz, @chennes, this pull request:

- Loads lightning settings when user opens _Preferences_ (yes, the settings were not loaded before this commit). Closes #15793
- Adds spinboxes to adjust the light direction using XYZ and orientation quaternion notations
- Removes the code that is supposed to adjust settings on-the-fly (it does not work anyway. If I make it work (by moving _View3DSettings_ variable into the class), _NaviCube_ shows up every time user presses the _Apply_ button. I could not find a way to make it go away). The code did not work because _View3DSettings_ object was created on the stack; in this case it is deleted after going out of its scope. 

- @adrianinsaval, this commit fixes showing up the dialog twice.

The code has been lightened, rewritten in _Qt_ way, now it is easier to read. Possibly, somebody, who understands _FreeCAD's_ allinternals, can make the viewer be adjusted when user changes settings, but I failed to make _NaviCube_ disappear. Perhaps, for now, this is just simplification of the source code and the UI.

I have tested this commit against Qt 5.15.14 and 6.7.2. The commit does not check if user enters an invalid value, though, I can't find harm in this, it does not seem to influence anything.

### CODE REVIEW

**Comment by chennes** [comment]
I think for the time-being (until Ubuntu 20.04 is no longer supported), you will need to explicitly create the vector for your calls to `setValueSilently`.

**Comment by xtemp09** [comment]
I completely forgot about older compilers; I'll rewrite some code fragments in old verbose way.

**Comment by FEA-eng** [comment]
So this will make it possible to reset the lighting in Preferences?

**Comment by xtemp09** [comment]
> So this will make it possible to reset the lighting in Preferences?


I didn't add a reset button. =/ One can set the direction to `(0, 0, -1)` or enter an invalid value.

**Comment by xtemp09** [comment]
If shaise hijacks _Coin_ and implements ambient occlusion, I can add 3-point lightning settings.

**Comment by FEA-eng** [comment]
> Well, I need some clarification. If a user presses _Reset page 'Light Sources'_, must the program load the direction of `(0, 0, -1)`? So far the program loads the direction stored in `user.cfg`.

It should load the default direction so the one that comes with clean FreeCAD config before changing anything in Preferences. It should be also changed to default when resetting at higher levels (preference group and all preferences instead of a page but I guess it’s linked).

**Comment by maxwxyz** [comment]
@xtemp09 Reset should load (0,0,-1) as this is the default setting. The preference reset should reset the values to the defaults and not the one currently in the config.

**Comment by xtemp09** [comment]
@maxwxyz, alright, perhaps, I slightly misunderstood the objective. Now if a user presses _Reset page 'Light Sources'..._ or chooses other reset options, it loads the default direction of `(0, 0, -1)`.

**Comment by FEA-eng** [comment]
I can confirm that resetting works as expected with this PR. It's a really great improvement, thanks.

**Comment by furgo16** [comment]
Related issue: https://github.com/FreeCAD/FreeCAD/issues/13591

**Comment by yorikvanhavre** [comment]
Although this has UI changes, we think it would be best to merge before 1.0.

> I completely forgot about older compilers; I'll rewrite some code fragments in old verbose way.

@xtemp09 is it good for merge already?

**Comment by yorikvanhavre** [comment]
@chennes I unilaterally decided to assign this to you as you were discussing it already ;)

**Comment by xtemp09** [comment]
> @xtemp09 is it good for merge already?


Yes. A couple of force pushes were made to make it compilable on _Ubuntu 20.04_ and to correct the tab order.

**Comment by xtemp09** [comment]
Now I conclude that this commit needs final touch to reach perfection. The `DlgSettingsLightSources.ui` contains tooltip which, I think, has to be more comprehensible.


@chennes, do you think this is better:

> Adjust the orientation of the directional light source by dragging the handle with the mouse or use the spin boxes for finer tuning.


?

**Comment by chennes** [comment]
> Now I conclude that this commit needs final touch to reach perfection. The `DlgSettingsLightSources.ui` contains tooltip which, I think, has to be more comprehensible.
> 
> @chennes, do you think this is better:
> 
> > Adjust the orientation of the directional light source by dragging the handle with the mouse or use the spin boxes for finer tuning.
> 
> ?

I'd use "fine tuning" rather than "finer tuning" but yes, overall I think that reaches the spirit of the operation correctly.

**Comment by FEA-eng** [comment]
@chennes Looks good to go now :-)

**Comment by FEA-eng** [comment]
@chennes Just to let you know, we discussed this during the merge meeting and agreed that it should go for 1.0: https://github.com/FreeCAD/FreeCAD/pull/15877#issuecomment-2296897492

File: src/Gui/PreferencePages/DlgSettingsLightSources.h [hunk 1]
```diff
@@ -47,29 +48,32 @@ class DlgSettingsLightSources : public PreferencePage
...
 
 private:
-    void setupConnection();
-    void toggleLight(bool on);
-    void lightIntensity(int value);
-    void lightColor();
     void saveDirection();
     void loadDirection();
-    QWidget* createViewer(QWidget* parent);
+    void createViewer(void);
```
Comment (chennes) [inline src/Gui/PreferencePages/DlgSettingsLightSources.h:70]:
```suggestion
    void createViewer();
```

File: src/Gui/PreferencePages/DlgSettingsLightSources.h [hunk 2]
```diff
@@ -47,29 +48,32 @@ class DlgSettingsLightSources : public PreferencePage
...
-    ~DlgSettingsLightSources() override;
+    ~DlgSettingsLightSources() override = default;
 
     void saveSettings() override;
     void loadSettings() override;
+    void resetSettingsToDefaults() override;
+
+public Q_SLOTS:
+    void updateDraggerQS (void);
+    void updateDraggerXYZ(void);
```
Comment (chennes) [inline src/Gui/PreferencePages/DlgSettingsLightSources.h:59]:
```suggestion
    void updateDraggerQS ();
    void updateDraggerXYZ();
```

**Review by chennes — APPROVED** [review]
No need to make these manually, I will merge them in a minute.

**Comment by maxwxyz** [comment]
@xtemp09 the dragger is now tiny and cannot be used interactively very easily, maybe because I have a HiDPI display?
This is on Windows:
grafik

**Comment by xtemp09** [comment]
@maxwxyz, there were conditions in which the arrow tip was outside of the viewport. I intentionally made the arrow smaller to avoid these situations.

**Comment by maxwxyz** [comment]
@xtemp09 ah ok, I thought the arrow is only an indicator as I usually dragged it at the other side. Is it possible to style this dragger, making the line/arrow smaller and maybe change the material (a darker one with more shininess would be better to see the actual light than the white dragger).

**Comment by xtemp09** [comment]
@maxwxyz, yes, the material of the dragger can be customized. [^1]


[^1]: https://www.coin3d.org/coin/classSoDirectionalLightDragger.html#pro-attribs

**Comment by maxwxyz** [comment]
I failed when I tried it the last time. Could you change it?
From https://github.com/FreeCAD/FreeCAD/issues/11166#issuecomment-1847401122
```
Values:
Ambient color: #333333
Diffuse color: #d2d2ff
Emissive color: #000000
Specular color: #cccccc
Shininess: 0.9
Transparency: 0
```

**Comment by xtemp09** [comment]
@maxwxyz, do you mean grey ball in dark grey viewport?

**Comment by maxwxyz** [comment]
@xtemp09 yes, shiny darker ball to better see the light direction as it would be in the 3D view. And there is no option to increase the size of the ball, decrease the arrow size or increase the height inside the preference window?

**Comment by xtemp09** [comment]
I think, to implement "sort of" zoom is easy by changing the position of the camera:


```cpp
auto cam = dynamic_cast <SoOrthographicCamera *> (view->getCamera());
cam->height = cam->height.getValue() * 2.0f;
```

**Comment by maxwxyz** [comment]
@xtemp09 do you plan on creating a PR with the above improvements and material changes?

**Comment by xtemp09** [comment]
@maxwxyz, you are suggesting quite a peculiar change in UI. Are you sure that designers and developers will agree with this?

**Comment by maxwxyz** [comment]
It's an improvement and on my screen it's practically not usable.

**Pull request merged**